src/net_processing.cpp

```diff
@@ -793,73 +724,35 @@ static void FindNextBlocksToDownload(NodeId nodeid, unsigned int count, std::vec
-    } else {
-        // Randomize the delay to avoid biasing some peers over others (such as due to
-        // fixed ordering of peer processing in ThreadMessageHandler)
-        process_time = last_request_time + GETDATA_TX_INTERVAL + GetRandMicros(MAX_GETDATA_RANDOM_DELAY);
-    }
-
-    // We delay processing announcements from inbound peers
-    if (use_inbound_delay) process_time += INBOUND_PEER_TX_DELAY;
-
-    // We delay processing announcements from peers that use txid-relay (instead of wtxid)
-    if (use_txid_delay) process_time += TXID_RELAY_DELAY;
-
-    return process_time;
-}
+} // namespace
 
-void RequestTx(CNodeState* state, const GenTxid& gtxid, std::chrono::microseconds current_time) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
+void PeerManager::RequestTx(const CNode& node, const GenTxid& gtxid, std::chrono::microseconds current_time)
 {
-    CNodeState::TxDownloadState& peer_download_state = state->m_tx_download;
-    if (peer_download_state.m_tx_announced.size() >= MAX_PEER_TX_ANNOUNCEMENTS ||
-            peer_download_state.m_tx_process_time.size() >= MAX_PEER_TX_ANNOUNCEMENTS ||
-            peer_download_state.m_tx_announced.count(gtxid.GetHash())) {
-        // Too many queued announcements from this peer, or we already have
-        // this announcement
+    AssertLockHeld(cs_main); // For m_txrequest
+    NodeId nodeid = node.GetId();
+    if (m_txrequest.CountTracked(nodeid) >= MAX_PEER_TX_ANNOUNCEMENTS) {
+        // Too many queued announcements from this peer
         return;
     }
-    peer_download_state.m_tx_announced.insert(gtxid.GetHash());
-
-    // Calculate the time to try requesting this transaction. Use
-    // fPreferredDownload as a proxy for outbound peers.
-    const auto process_time = CalculateTxGetDataTime(gtxid, current_time, !state->fPreferredDownload, !state->m_wtxid_relay && g_wtxid_relay_peers > 0);
-
-    peer_download_state.m_tx_process_time.emplace(process_time, gtxid);
+    auto state = State(nodeid);
+
+    // Decide the TxRequestTracker parameters for this announcement:
+    // - "overloaded": if at least MAX_PEER_TX_LOAD requests are in flight and requestable combined.
+    // - "preferred": if fPreferredDownload is set (= outbound, or PF_NOBAN permission)
+    // - "reqtime": immediately if none of the penalties below apply, otherwise current time plus:
+    //   - INBOUND_PEER_TX_DELAY for announcements from non-preferred connections
+    //   - TXID_RELAY_DELAY for announcements from txid peers while wtxid peers are available
+    //   - OVERLOADED_PEER_TX_DELAY for announcements from overloaded peers
+    auto delay = std::chrono::microseconds{0};
+    bool preferred = state->fPreferredDownload;
+    bool overloaded = m_txrequest.CountLoad(nodeid) >= MAX_PEER_TX_LOAD;
+    if (!preferred) delay += INBOUND_PEER_TX_DELAY;
+    if (!state->m_wtxid_relay && g_wtxid_relay_peers > 0) delay += TXID_RELAY_DELAY;
+    if (overloaded) delay += OVERLOADED_PEER_TX_DELAY;
```

Comment:
I wonder if it wouldn't make more sense to move the `overloaded` judgement entirely into `ReceivedInv` and instead pass the peer's `MAX_PEER_TX_LOAD` in and have `OVERLOADED_PEER_TX_DELAY` passed in when constructing`m_txrequest`. So `ReceivedInv` would then do `if (CountLoad(peer) > max_load) { first = false; reqtime += m_overloaded_delay; }`.

The reason I say that is I think the "first" marker logic doesn't really make sense if you don't bump the reqtime for overloaded peers -- you'd end up with the overloaded peer being selected while the first non-overloaded peer was still waiting for its reqtime to arrive.

The `!preferred` delay could also be handled the same way. I guess you'd need to pass in `current_time` and `peer_delay = TXID_RELAY_DELAY` (when appropriate) (so to speak) for that work. Again, there needs to be a positive delay for `!preferred` peers or the logic won't work right, and while it can vary amongst announcements in limited ways without breaking things, I think it would be hard to do that correctly, so being set at the txrequest level wouldn't be a big problem.

That might also have the advantage of being able to pass in `delay == 0` rather than `reqtime == microseconds::min()` which seems a little bit more obvious.

Reply:
I did this at first, but I like the current approach a lot more. It leaves the actual policy decisions inside net_processing, without needing configuration flags etc. in the txrequest interface.

Not a very strong opinion, but I like the fact that txrequest is mostly the data structure, and its responsibility is consistency and efficiency - not policy parameters. There are currently a lot of future improvements that can be made without changing txrequest - and that due to the nature of its current interface - are already covered by fuzz testing.

Reply:
> There are currently a lot of future improvements that can be made without changing txrequest - and that due to the nature of its current interface are already covered by fuzz testing.

Kind-of? I would have said the fuzz testing only really tests the optimisations work; it doesn't really check that the basic implementation makes "sense"?

eg, we could check the "don't request dependent transactions out of order" goal by adding a constraint that all peers announce tx `i` prior to tx `j`, and then checking that tx `i` is requested before tx `j`, but I think that would currently fail in the fuzzer due to random delays (and also fail in net_processing in rare cases, eg where a tx is only announced by one node, `j` is announced less than 2 seconds after `i` and the node becomes non-overloaded in that same period, or where all the non-wtxid-relay peers disappear in between).

I'm trying out some logging to see if the "first" marker actually does much good in practice. We're calling `GetRequestable`/`RequestedTx` ten times a second for every peer (provided various queues don't fill up anyway), so I think you'd have to be pretty unlucky for it to even matter.

```diff
--- a/src/txrequest.cpp
+++ b/src/txrequest.cpp
@@ -528,6 +528,29 @@ public:
         // in between, which preserve the state of other GenTxids).
         assert(it != m_index.get<ByPeer>().end());
         assert(it->GetState() == State::CANDIDATE_BEST);
+       {
+            auto bytxhashit = m_index.project<ByTxHash>(it);
+            auto bytxhashit_pre = bytxhashit;
+           int delayed = 0;
+           int candidates = 0;
+           while (bytxhashit_pre != m_index.get<ByTxHash>().begin()) {
+               --bytxhashit_pre;
+               if (bytxhashit_pre->GetState() != State::CANDIDATE_DELAYED || bytxhashit->m_txhash != it->m_txhash) break;
+               ++delayed;
+           }
+           ++bytxhashit;
+           while (bytxhashit != m_index.get<ByTxHash>().end() && bytxhashit->GetState() != State::COMPLETED && bytxhashit->m_txhash == it->m_txhash) {
+               ++candidates;
+               ++bytxhashit;
+           }
+           int completed = 0;
+           while (bytxhashit != m_index.get<ByTxHash>().end() && bytxhashit->GetState() == State::COMPLETED && bytxhashit->m_txhash == it->m_txhash) {
+               ++completed;
+               ++bytxhashit;
+           }
+           assert(bytxhashit == m_index.get<ByTxHash>().end() || bytxhashit->m_txhash != it->m_txhash);
+           LogPrintf("ABCD requested txid=%s preferred=%d first=%d delayed=%d candidates=%d completed=%d peer=%d\n", it->m_txhash.ToString(), it->m_preferred, it->m_first, delayed, candidates, completed, it->m_peer);
+       }
         Modify<ByPeer>(it, [expiry](Entry& entry) {
             entry.SetState(State::REQUESTED);
             entry.m_time = expiry;
```

After running it for a few minutes with ~20 peers, only 0.17% of requests actually had other announcements in CANDIDATE_READY, so for 99.82% of txs the first marker didn't make any difference. About 1.6% of requests weren't for a "first" marked announcement (a tor outbound managed to overload itself I think). Will let it go 'til it fills up the disk with logs or whatever, and see what it looks like then.

Reply:
> Kind-of? I would have said the fuzz testing only really tests the optimisations work; it doesn't really check that the basic implementation makes "sense"?

Yeah, kind of.

The fuzz test only verifies consistency of the optimized data structure, and observational equivalence with the naive reimplementation. That naive reimplementation is hopefully close enough to the written specification in txrequest.h that we can say the test actually verifies the specification. It however does not test whether that specification actually results in desirable high-level behavior. That's what the scenario unit tests are for, but they're not nearly as detailed.

Still, as far as the specification goes, by keeping the policy out of txrequest and it tests, there is confidence that the data structure correctly implements it, and this doesn't depend on coincidences like "non-preferred requests always have a reqtime in the future" that are currently true, but maybe won't always be true. The interface doesn't force such relations, so the fuzz tests should be covering both situations where they do and don't hold.

Furthermore, I'm not sure there is much of a simplification in the interface that can be made by moving more into txrequest?
* Right now: `ReceivedInv(peer, gtxid, preferred, overloaded, reqtime)`.
* Potentially: `ReceivedInv(peer, gtxid, preferred, have_wtxid_peers, current_time)`.
* With the addition of bypassing `overloaded` under PF_RELAY you'd need to pass an additional argument `pf_relay_peer` argument even.

The fact that this last change could be made by just changing the parameters to `ReceivedInv` and not changing the interface - or the fuzz tests - is sign in favor of it, IMO.

Several potential changes like making delays dynamic in function of load/inflight, or adding randomness to it, are equally possible without impacting txrequest.

Reply:
Yeah, agree with what you wrote. (Well, I was thinking you'd call `ReceivedInv(peer, gtxid, preferred, max_load, current_time)` and pass `max_load=max()` to bypass overloaded, fwiw)

I'm more thinking of the API in terms of what the caller has to do to ensure the txrequest modules behaves "sensibly" -- avoids requesting txs out of order, avoiding double requesting a tx, etc. If there's a way to make txrequest behave "sensibly" no matter how you call the API, that might be better than just documenting the expectations, or the consequences of violating the expectations. But I don't fully understand what sensibly means or what restrictions that implies yet.

Reply:
I see what you mean now.

There is no strict reason why we couldn't have both. There could a class around TxRequestTracker that adds the policy and sanity checking of inputs, which just forwards to the inner TxRequestTracker. That would let the inner part still be tested for consistency, while permitting testing of the higher level construct for higher level "goal" properties.

As the logic in that outer layer would just be (part of) what is now in TxRequest(), I don't think this is worth it - but maybe at some point it is.

Reply:
Yeah. I don't think a wrapper would really be an improvement on documenting the expectations, and agree that sticking fixed params in txrequest is really too restrictive. Going to mark this thread as resolved, but still continue thinking about it.